jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java

```diff
@@ -306,36 +311,61 @@ private void obtainBaseImagesLayers(
                       "scheduling obtaining base images layers",
                       results.baseImagesAndRegistryClient.get().images.size())) {
 
+                Map<DescriptorDigest, Future<PreparedLayer>> preparedLayersCache = new HashMap<>();
                 Map<Image, List<Future<PreparedLayer>>> baseImagesAndLayers = new HashMap<>();
                 for (Image baseImage : results.baseImagesAndRegistryClient.get().images) {
                   List<Future<PreparedLayer>> layers =
                       obtainBaseImageLayers(
-                          baseImage, layersRequiredLocally, progressDispatcher.newChildProducer());
+                          baseImage,
+                          layersRequiredLocally,
+                          preparedLayersCache,
+                          progressDispatcher.newChildProducer());
                   baseImagesAndLayers.put(baseImage, layers);
                 }
                 return baseImagesAndLayers;
               }
             });
   }
 
-  private List<Future<PreparedLayer>> obtainBaseImageLayers(
+  @VisibleForTesting
+  List<Future<PreparedLayer>> obtainBaseImageLayers(
```

Comment:
I thought that maybe this could have threading issues, but we never call this in parallel right? Is there any value in marking this not threadsafe?

Reply:
Added a comment
```
// This method updates the given "preparedLayersCache" and should not be called concurrently.
```
. (FYI, `@NotThreadSafe` can only be applied to a class.)

The non-blocking method is called only by a single thread task inside `obtainBaseImagesLayers()`.